### PR TITLE
Changes from MAML to OML

### PR DESCRIPTION
**Pull request by lucasosouza**
Some rough ideas of how to move from MAML to OML, just to make the discussion more concrete.

### CODE REVIEW

File: examples/vision/oml_omniglot.py [hunk 1]
```diff
@@ -37,15 +37,25 @@ def fast_adapt(batch, learner, loss, adaptation_steps, shots, ways, device):
...
 
     # Evaluate the adapted model
     predictions = learner(evaluation_data)
     valid_error = loss(predictions, evaluation_labels)
     valid_error /= len(evaluation_data)
     valid_accuracy = accuracy(predictions, evaluation_labels)
-    return valid_error, valid_accuracy
+    return valid_error, valid_accuracy, evaluation_data, evaluation_labels
 
+def sample_random_set(batch, shots, ways, device):
```
Comment (lucasosouza) [inline examples/vision/oml_omniglot.py:49]: Random set should sample from all tasks? How much samples exactly are sampled in OML?

File: examples/vision/oml_omniglot.py [hunk 2]
```diff
@@ -89,31 +99,41 @@ def main(
         meta_train_accuracy = 0.0
         meta_valid_error = 0.0
         meta_valid_accuracy = 0.0
+
+        learner = maml.clone()
```
Comment (lucasosouza) [inline examples/vision/oml_omniglot.py:103]: Moved clone outside the task loop.

File: examples/vision/oml_omniglot.py [hunk 3]
```diff
@@ -89,31 +99,41 @@ def main(
...
+        evaluation_data_all_tasks = []
+        evaluation_labels_all_tasks = []
```
Comment (lucasosouza) [inline examples/vision/oml_omniglot.py:105]: For each task we need to gather samples for the outer loop update. Here i am using the same data used in the inner loop to calculate the validation accuracy. Since it is not the same data used for training, it should have the same effect.

File: examples/vision/oml_omniglot.py [hunk 4]
```diff
@@ -89,31 +99,41 @@ def main(
...
-            evaluation_error.backward()
+            evaluation_error, evaluation_accuracy, evaluation_data, evaluation_labels = \
+                fast_adapt(batch,
+                          learner,
+                          loss,
+                          adaptation_steps,
+                          shots,
+                          ways,
+                          device,
+                          freeze_feature_extraction=True)
```
Comment (lucasosouza) [inline examples/vision/oml_omniglot.py:118]: The freeze feature extraction is what guarantees that $\theta$ is not updated in the inner loop

File: examples/vision/oml_omniglot.py [hunk 5]
```diff
@@ -125,9 +145,18 @@ def main(
...
-        for p in maml.parameters():
-            p.grad.data.mul_(1.0 / meta_batch_size)
+        # sample remember/random set
+        random_batch = tasksets.train.sample()
+        random_data, random_labels = sample_random_set(
+            random_batch, shots, ways, device
+        )
+        # concatenate with trajectory
+        meta_training_test_data = torch.stack(evaluation_data_all_tasks, random_data)
+        meta_training_test_labels = torch.stack(evaluation_labels_all_tasks, random_labels)
```
Comment (lucasosouza) [inline examples/vision/oml_omniglot.py:155]: Concatenates the random set, sampled from all tasks, with samples from the trajectory. Notice the samples are not the same used in inner loop.

File: examples/vision/oml_omniglot.py [hunk 6]
```diff
@@ -125,9 +145,18 @@ def main(
...
+
+        # learn - outer loop
+        loss = loss(learner(meta_training_test_data), meta_training_test_labels)
+        loss.backward()
```
Comment (lucasosouza) [inline examples/vision/oml_omniglot.py:159]: Here is the only time `backward` is called, in the outer loop update. Can't be called in the inner loop, since the computational graph points to the original weights - hence the update is done manually through the `maml_update` function.

I'm assuming the entire network is updated in the outer loop, including the feature extractor (theta) and the adaptation params (W). That is something we still have to validate.

File: learn2learn/vision/models.py [hunk 7]
```diff
@@ -168,6 +168,28 @@ def forward(self, x):
...
+
+    def __init__(self, input_size, output_size, sizes=None):
+        if sizes is None:
+            sizes = [256, 128, 64, 64]
+        layers = [LinearBlock(input_size, sizes[0]), ]
+        for s_i, s_o in zip(sizes[:-1], sizes[1:]):
+            layers.append(LinearBlock(s_i, s_o))
+        # break down into features and adaptaion
+        self.features = nn.Sequential(*layers)
+        self.adaptation = fc_init_(nn.Linear(sizes[-1], output_size))
```
Comment (lucasosouza) [inline learn2learn/vision/models.py:184]: This network separates adaptation params from the feature extractor, making it easier to freeze later.

File: examples/vision/oml_omniglot.py [hunk 8]
```diff
@@ -89,31 +99,41 @@ def main(
...
-            evaluation_error, evaluation_accuracy = fast_adapt(batch,
-                                                               learner,
-                                                               loss,
-                                                               adaptation_steps,
-                                                               shots,
-                                                               ways,
-                                                               device)
-            evaluation_error.backward()
+            evaluation_error, evaluation_accuracy, evaluation_data, evaluation_labels = \
+                fast_adapt(batch,
```
Comment (lucasosouza) [inline examples/vision/oml_omniglot.py:111]: What is shot and ways, and how can we understand these parameters in the context of OML?

File: learn2learn/algorithms/maml.py [hunk 9]
```diff
@@ -135,9 +140,14 @@ def adapt(self,
             allow_nograd = self.allow_nograd
         second_order = not first_order
 
+        if freeze_feature_extraction:
```
Comment (lucasosouza) [inline learn2learn/algorithms/maml.py:143]: `freeze_feature_extraction` guarantees only the adaptation params are updated.

File: learn2learn/vision/models.py (hunk 7 above)
Comment (lucasosouza) [reply learn2learn/vision/models.py:184]: A cleaner way to make this extensible to more models is to imbue the model with a `get_adaptation_params` function, and on the experiment side we check if the given model has this function in the setup before proceeding.

That should also make it easier to test different versions of the model where more or less layers are frozen during the inner loop, as it is done in the OML and ANML papers.

File: examples/vision/oml_omniglot.py (hunk 2 above)
Comment (lucasosouza) [reply examples/vision/oml_omniglot.py:103]: To not clone the entire network and make the representation/adaptation difference explicit, we can do something like this
in the inner loop:
```
representation_net = self.model.representation
adaptation_net = self.model.adaptation.clone()
# inside the inner loop
output = adaptation_net(representation_net(x))
...
inner_step(adaptation_net) # only updates the adaptation net, without calling backward 
```
in the outer loop:
```
output = adaptation_net(representation_net(x))
...
loss.backward()
opt.step() # updates the entire network
```
